### ISSUE TITLE
Underflow error on view (follow up to issue #33)

### ISSUE THREAD

**Issue by fils** [rana/ora#36]
Tamás,
  I was playing more with the updated ora based on issue #33.   It is working for me but I can get myself to another underflow error.  

  If I create a view on the table in the test code with
`
if _, err := testDb.Exec(`CREATE VIEW test_janus_v AS SELECT * FROM test_janus`); err != nil {
		t.Fatal(err)
	}
​`
that will PASS the test.,

However, we found that there are views (which seem to be composites from several tables) that do not pass.   If we then make a table from this view that fails, that will in fact pass the test (not underflow).

We think we can build a test for you that shows this, but it will take time to build the tables, data and view to do so.   I don't know if I can get it done this weekend or not.   I will try.

I wanted to start this issue though in case the information on an underflow error from a view  was enough for you to start tracking things down like you did that resolved it on the table level.    

Thanks..  I will be in touch

**Comment by tgulacsi** [comment]
Can you tell us the results (column types) of what github.com/tgulacsi/go/orahlp.DescribeSQLQuery returns? Maybe both about the composite query and about the created table - maybe there we can spot the difference...

It should be the same as what user_tab_cols contains about the columns, but I don't know.
I'd check that about the used tables, too.

**Comment by fils** [comment]
Tamás, my DBA and I will try and collect what we can post to this thread.  If we can get a good test together will pop it in a gist and post it as well.   Thanks for the guidance on some things to collect.

**Comment by divdat** [comment]
Here is the information about the table and the view from all_tab_columns in the database:

Data from all_tab_columns

Table:  ocd_chem_carb_table (created from the view)
TABLE_NAME||':'||COLUMN_NAME||':'||DATA_TYPE||':'||DATA_LENGTH||':'||DATA_PRECISION||':'||DATA_SCALE||':'||NULLABLE
------------------------------------------------------------------------------------------------------------------------
OCD_CHEM_CARB_TABLE:LEG:NUMBER:22:5:0:N
OCD_CHEM_CARB_TABLE:SITE:NUMBER:22:6:0:N
OCD_CHEM_CARB_TABLE:HOLE:VARCHAR2:1:::N
OCD_CHEM_CARB_TABLE:CORE:NUMBER:22:5:0:N
OCD_CHEM_CARB_TABLE:CORE_TYPE:VARCHAR2:1:::N
OCD_CHEM_CARB_TABLE:SECTION_NUMBER:NUMBER:22:2:0:N
OCD_CHEM_CARB_TABLE:SECTION_TYPE:VARCHAR2:2:::Y
OCD_CHEM_CARB_TABLE:TOP_CM:NUMBER:22:::Y
OCD_CHEM_CARB_TABLE:BOT_CM:NUMBER:22:::Y
OCD_CHEM_CARB_TABLE:DEPTH_MBSF:NUMBER:22:::Y
OCD_CHEM_CARB_TABLE:INOR_C_WT_PCT:NUMBER:22:::Y
OCD_CHEM_CARB_TABLE:CACO3_WT_PCT:NUMBER:22:::Y
OCD_CHEM_CARB_TABLE:TOT_C_WT_PCT:NUMBER:22:::Y
OCD_CHEM_CARB_TABLE:ORG_C_WT_PCT:NUMBER:22:::Y
OCD_CHEM_CARB_TABLE:NIT_WT_PCT:NUMBER:22:::Y
OCD_CHEM_CARB_TABLE:SUL_WT_PCT:NUMBER:22:::Y
OCD_CHEM_CARB_TABLE:H_WT_PCT:NUMBER:22:::Y


View:  ocd_chem_carb_v (created from a query)
TABLE_NAME||':'||COLUMN_NAME||':'||DATA_TYPE||':'||DATA_LENGTH||':'||DATA_PRECISION||':'||DATA_SCALE||':'||NULLABLE
------------------------------------------------------------------------------------------------------------------------
OCD_CHEM_CARB_V:LEG:NUMBER:22:5:0:N
OCD_CHEM_CARB_V:SITE:NUMBER:22:6:0:N
OCD_CHEM_CARB_V:HOLE:VARCHAR2:1:::N
OCD_CHEM_CARB_V:CORE:NUMBER:22:5:0:N
OCD_CHEM_CARB_V:CORE_TYPE:VARCHAR2:1:::N
OCD_CHEM_CARB_V:SECTION_NUMBER:NUMBER:22:2:0:N
OCD_CHEM_CARB_V:SECTION_TYPE:VARCHAR2:2:::Y
OCD_CHEM_CARB_V:TOP_CM:NUMBER:22:::Y
OCD_CHEM_CARB_V:BOT_CM:NUMBER:22:::Y
OCD_CHEM_CARB_V:DEPTH_MBSF:NUMBER:22:::Y
OCD_CHEM_CARB_V:INOR_C_WT_PCT:NUMBER:22:::Y
OCD_CHEM_CARB_V:CACO3_WT_PCT:NUMBER:22:::Y
OCD_CHEM_CARB_V:TOT_C_WT_PCT:NUMBER:22:::Y
OCD_CHEM_CARB_V:ORG_C_WT_PCT:NUMBER:22:::Y
OCD_CHEM_CARB_V:NIT_WT_PCT:NUMBER:22:::Y
OCD_CHEM_CARB_V:SUL_WT_PCT:NUMBER:22:::Y
OCD_CHEM_CARB_V:H_WT_PCT:NUMBER:22:::Y

There is no difference between the view created from a query, and the table created from the view.

**Comment by fils** [comment]
It's a bit messy...

```
Doug-Fils-MBP:ocdGarden dfils$ go test
2015/09/11 09:50:35 Ready to connect
2015/09/11 09:50:35 Connected
2015/09/11 09:50:38 desc: []orahlp.Column{orahlp.Column{Schema:"", Name:"LEG", Type:2, Length:22, Precision:5, Scale:0, Nullable:false, CharsetID:0, CharsetForm:0}, orahlp.Column{Schema:"", Name:"SITE", Type:2, Length:22, Precision:6, Scale:0, Nullable:false, CharsetID:0, CharsetForm:0}, orahlp.Column{Schema:"", Name:"HOLE", Type:1, Length:1, Precision:0, Scale:0, Nullable:false, CharsetID:873, CharsetForm:1}, orahlp.Column{Schema:"", Name:"CORE", Type:2, Length:22, Precision:5, Scale:0, Nullable:false, CharsetID:0, CharsetForm:0}, orahlp.Column{Schema:"", Name:"CORE_TYPE", Type:1, Length:1, Precision:0, Scale:0, Nullable:false, CharsetID:873, CharsetForm:1}, orahlp.Column{Schema:"", Name:"SECTION_NUMBER", Type:2, Length:22, Precision:2, Scale:0, Nullable:false, CharsetID:0, CharsetForm:0}, orahlp.Column{Schema:"", Name:"SECTION_TYPE", Type:1, Length:2, Precision:0, Scale:0, Nullable:true, CharsetID:873, CharsetForm:1}, orahlp.Column{Schema:"", Name:"TOP_CM", Type:2, Length:22, Precision:0, Scale:0, Nullable:true, CharsetID:0, CharsetForm:0}, orahlp.Column{Schema:"", Name:"BOT_CM", Type:2, Length:22, Precision:0, Scale:0, Nullable:true, CharsetID:0, CharsetForm:0}, orahlp.Column{Schema:"", Name:"INOR_C_WT_PCT", Type:2, Length:22, Precision:0, Scale:0, Nullable:true, CharsetID:0, CharsetForm:0}, orahlp.Column{Schema:"", Name:"CACO3_WT_PCT", Type:2, Length:22, Precision:0, Scale:0, Nullable:true, CharsetID:0, CharsetForm:0}, orahlp.Column{Schema:"", Name:"TOT_C_WT_PCT", Type:2, Length:22, Precision:0, Scale:0, Nullable:true, CharsetID:0, CharsetForm:0}, orahlp.Column{Schema:"", Name:"ORG_C_WT_PCT", Type:2, Length:22, Precision:0, Scale:0, Nullable:true, CharsetID:0, CharsetForm:0}, orahlp.Column{Schema:"", Name:"NIT_WT_PCT", Type:2, Length:22, Precision:0, Scale:0, Nullable:true, CharsetID:0, CharsetForm:0}, orahlp.Column{Schema:"", Name:"SUL_WT_PCT", Type:2, Length:22, Precision:0, Scale:0, Nullable:true, CharsetID:0, CharsetForm:0}, orahlp.Column{Schema:"", Name:"H_WT_PCT", Type:2, Length:22, Precision:0, Scale:0, Nullable:true, CharsetID:0, CharsetForm:0}}
2015/09/11 09:50:38 Results: 171 1049 B 3 B 4 @ {0 false} {6.12 true} {7.12 true} {8 true} {9.99 true} {0 false} {11 true} {0 false} {0.8 true} {0.42 true}
--- FAIL: TestFils (3.25s)
	janusNillIssue_test.go:232: Env.ociError OCI-22054: underflow error

FAIL
exit status 1
FAIL	opencoredata.org/ocdGarden	5.677s

```

**Comment by fils** [comment]
I did the wrong tables, views I think in the previous one,   let me try again

for ocd_chem_carb_table

```
desc: []orahlp.Column{orahlp.Column{Schema:"", Name:"LEG", Type:2, Length:22, Precision:5, Scale:0, Nullable:false, CharsetID:0, CharsetForm:0}, orahlp.Column{Schema:"", Name:"SITE", Type:2, Length:22, Precision:6, Scale:0, Nullable:false, CharsetID:0, CharsetForm:0}, orahlp.Column{Schema:"", Name:"HOLE", Type:1, Length:1, Precision:0, Scale:0, Nullable:false, CharsetID:873, CharsetForm:1}, orahlp.Column{Schema:"", Name:"CORE", Type:2, Length:22, Precision:5, Scale:0, Nullable:false, CharsetID:0, CharsetForm:0}, orahlp.Column{Schema:"", Name:"CORE_TYPE", Type:1, Length:1, Precision:0, Scale:0, Nullable:false, CharsetID:873, CharsetForm:1}, orahlp.Column{Schema:"", Name:"SECTION_NUMBER", Type:2, Length:22, Precision:2, Scale:0, Nullable:false, CharsetID:0, CharsetForm:0}, orahlp.Column{Schema:"", Name:"SECTION_TYPE", Type:1, Length:2, Precision:0, Scale:0, Nullable:true, CharsetID:873, CharsetForm:1}, orahlp.Column{Schema:"", Name:"TOP_CM", Type:2, Length:22, Precision:0, Scale:-127, Nullable:true, CharsetID:0, CharsetForm:0}, orahlp.Column{Schema:"", Name:"BOT_CM", Type:2, Length:22, Precision:0, Scale:-127, Nullable:true, CharsetID:0, CharsetForm:0}, orahlp.Column{Schema:"", Name:"INOR_C_WT_PCT", Type:2, Length:22, Precision:0, Scale:-127, Nullable:true, CharsetID:0, CharsetForm:0}, orahlp.Column{Schema:"", Name:"CACO3_WT_PCT", Type:2, Length:22, Precision:0, Scale:-127, Nullable:true, CharsetID:0, CharsetForm:0}, orahlp.Column{Schema:"", Name:"TOT_C_WT_PCT", Type:2, Length:22, Precision:0, Scale:-127, Nullable:true, CharsetID:0, CharsetForm:0}, orahlp.Column{Schema:"", Name:"ORG_C_WT_PCT", Type:2, Length:22, Precision:0, Scale:-127, Nullable:true, CharsetID:0, CharsetForm:0}, orahlp.Column{Schema:"", Name:"NIT_WT_PCT", Type:2, Length:22, Precision:0, Scale:-127, Nullable:true, CharsetID:0, CharsetForm:0}, orahlp.Column{Schema:"", Name:"SUL_WT_PCT", Type:2, Length:22, Precision:0, Scale:-127, Nullable:true, CharsetID:0, CharsetForm:0}, orahlp.Column{Schema:"", Name:"H_WT_PCT", Type:2, Length:22, Precision:0, Scale:-127, Nullable:true, CharsetID:0, CharsetForm:0}}
```


for ocd_chem_carb_v

```
desc: []orahlp.Column{orahlp.Column{Schema:"", Name:"LEG", Type:2, Length:22, Precision:5, Scale:0, Nullable:false, CharsetID:0, CharsetForm:0}, orahlp.Column{Schema:"", Name:"SITE", Type:2, Length:22, Precision:6, Scale:0, Nullable:false, CharsetID:0, CharsetForm:0}, orahlp.Column{Schema:"", Name:"HOLE", Type:1, Length:1, Precision:0, Scale:0, Nullable:false, CharsetID:873, CharsetForm:1}, orahlp.Column{Schema:"", Name:"CORE", Type:2, Length:22, Precision:5, Scale:0, Nullable:false, CharsetID:0, CharsetForm:0}, orahlp.Column{Schema:"", Name:"CORE_TYPE", Type:1, Length:1, Precision:0, Scale:0, Nullable:false, CharsetID:873, CharsetForm:1}, orahlp.Column{Schema:"", Name:"SECTION_NUMBER", Type:2, Length:22, Precision:2, Scale:0, Nullable:false, CharsetID:0, CharsetForm:0}, orahlp.Column{Schema:"", Name:"SECTION_TYPE", Type:1, Length:2, Precision:0, Scale:0, Nullable:true, CharsetID:873, CharsetForm:1}, orahlp.Column{Schema:"", Name:"TOP_CM", Type:2, Length:22, Precision:0, Scale:0, Nullable:true, CharsetID:0, CharsetForm:0}, orahlp.Column{Schema:"", Name:"BOT_CM", Type:2, Length:22, Precision:0, Scale:0, Nullable:true, CharsetID:0, CharsetForm:0}, orahlp.Column{Schema:"", Name:"INOR_C_WT_PCT", Type:2, Length:22, Precision:0, Scale:0, Nullable:true, CharsetID:0, CharsetForm:0}, orahlp.Column{Schema:"", Name:"CACO3_WT_PCT", Type:2, Length:22, Precision:0, Scale:0, Nullable:true, CharsetID:0, CharsetForm:0}, orahlp.Column{Schema:"", Name:"TOT_C_WT_PCT", Type:2, Length:22, Precision:0, Scale:0, Nullable:true, CharsetID:0, CharsetForm:0}, orahlp.Column{Schema:"", Name:"ORG_C_WT_PCT", Type:2, Length:22, Precision:0, Scale:0, Nullable:true, CharsetID:0, CharsetForm:0}, orahlp.Column{Schema:"", Name:"NIT_WT_PCT", Type:2, Length:22, Precision:0, Scale:0, Nullable:true, CharsetID:0, CharsetForm:0}, orahlp.Column{Schema:"", Name:"SUL_WT_PCT", Type:2, Length:22, Precision:0, Scale:0, Nullable:true, CharsetID:0, CharsetForm:0}, orahlp.Column{Schema:"", Name:"H_WT_PCT", Type:2, Length:22, Precision:0, Scale:0, Nullable:true, CharsetID:0, CharsetForm:0}}
```

**Comment by divdat** [comment]
From the output from describing the query with orahlp:

The "scale" values for the NUMBER fields in the table are -127
The "scale" values for the NUMBER fields in the view are 0

That must be why the underflow from views ...

**Comment by tgulacsi** [comment]
I agree. The code which decides whether to use float or int as destination,
chooses baded on precision and scale. See my last commit, that changed that.
So more tweaks needed there...
2015. szept. 11. du. 5:17 ezt írta ("Patrick Diver" <
notifications@github.com>):

**Comment by divdat** [comment]
One further note -- Doug ran the orahlp on the "big query" itself, the one that is used to create the view, and it also is coming back with scale = 0 on the number fields ...

**Comment by divdat** [comment]
I was looking at the fix for issue #33, line 349-350:

if precision == 0 && numericScale == -127 {
	gct = rset.stmt.cfg.Rset.numberFloat

perhaps it should be

if precision == 0 && (numericScale == 0 || numericScale == -127) {
        gct = rset.stmt.cfg.Rset.numberFloat

...  just an idea ...

**Comment by tgulacsi** [comment]
Exactly. I'll try it tomorrow.
2015. szept. 11. du. 6:34 ezt írta ("Patrick Diver" <
notifications@github.com>):

**Comment by tgulacsi** [comment]
Ok, I can't reproduce it, so just pushed it. Please test the master branch.

---
https://onename.com/tgulacsi

2015-09-11 19:27 GMT+02:00 Gulácsi Tamás <gthomas@gthomas.eu>:

**Comment by divdat** [comment]
We tested it, and everything is underflowing now ... even what worked before, so that didn't work!

**Comment by tgulacsi** [comment]
Ok, reverted it.

Can you help in reproducing the error?

Start with a table with minimal number of fields (one?) that reproduces the
error,
then tell us the CREATE TABLE, CREATE VIEW. I assume that the underflow
happens when the field contains a float between 0.1 and 0.9...

---
https://onename.com/tgulacsi

2015-09-11 23:01 GMT+02:00 Patrick Diver <notifications@github.com>:

**Comment by rana** [comment]
Hi Douglas,

Here's a workaround, if not a fix. To correct the issue define each table column with a scale if you wish to use sql.NullFloat64. For example, 

```sql
CREATE TABLE test_janus (
		leg NUMBER(5),
		site NUMBER(6),
		hole VARCHAR2(1),
		core NUMBER(5),
		core_type VARCHAR2(1),
		section_number NUMBER(2),
		section_Type VARCHAR2(2) NULL,
		top_cm NUMBER(6,3) NULL,
		bot_cm NUMBER(6,3) NULL,
		depth_mbsf NUMBER(6,3) NULL,
		inor_c_wt_pct NUMBER(6,3) NULL,
		caco3_wt_pct NUMBER(6,3) NULL,
		tot_c_wt_pct NUMBER(6,3) NULL,
		org_c_wt_pct NUMBER(6,3) NULL,
		nit_wt_pct NUMBER(6,3) NULL,
		sul_wt_pct NUMBER(6,3) NULL,
		h_wt_pct NUMBER(6,3) NULL
	)
```

Defining a column without specifying the scale e.g., `inor_c_wt_pct NUMBER NULL`, sets the scale to zero. The ora driver sees that the scale is zero and infers that the correct Go DataType is an integer and attempts to load an int64. If your intention is to use sql.NullInt64, don't insert floating points into that column. 

Oracle's [NUMBER](http://docs.oracle.com/cd/B28359_01/server.111/b28318/datatype.htm#CNCPT1832) documentation says "If a precision is not specified, the column stores values as given. If no scale is specified, the scale is zero." Even though Oracle documents the [NUMBER definition](http://docs.oracle.com/cd/B28359_01/server.111/b28318/datatype.htm#CNCPT1832) to have a zero scale when no scale is specified, Oracle appears to allow floating point numbers anyway, as your example shows. I believe we are in an undocumented Oracle gray area which solved simply by specifying a scale or sql.NullInt64 Go type. If there is a driver fix please specify or check in.

Let me know if you have any questions.

Thank you,

Rana

---
[Your test working with correction](https://gist.github.com/rana/b9d2edd9d12fe902b958)

**Comment by tgulacsi** [comment]
@divdat @fils 
I've tried to reproduce, without success. Obviously my environment (DB & table & view) is not the same as yours.

On "issue36" branch, I've refactored the logic onto the end of rsetCfg.go, and added a test in rsetCfg_test.go, so it is easier to change and test it.

If you still can reproduce the problem, then please:
1. document the DB version, the CREATE TABLE, CREATE VIEW and the used SELECT statements.
2. try to reproduce it by tweaking TestUnderflow in z_oracle_test.go
3. tweak numericColumnType in resetCfg.go to fix your issue, or at least document the prefision&scale of the ominous column.

Thanks in advance!

**Comment by divdat** [comment]
Doug will be testing, I need to get my Mac updated to 10.8 before I can install Go ...

Just a note:  I have tested using cx_Oracle in Python and I do not get the underflow issue.  This is for
testing against the raw query, the synonym of the view created from the query, and also from a table created from the view.  Simply connecting, setting up a cursor, executing, and printing the results.

I also have been able to use cx_Oracle with SQLAlchemy and Pandas and it also works.



I noticed in your tests that you do not have any numbers created into the tests between 0.1 and 0.9 which I think is what is creating the problem.

**Comment by divdat** [comment]
Here is the cx_Oracle cursor description for each method of retrieving the data:

------------------- raw query
[('LEG', <type 'cx_Oracle.NUMBER'>, 6, 22, 5, 0, 0), ('SITE', <type 'cx_Oracle.NUMBER'>, 7, 22, 6, 0, 0), ('HOLE', <type 'cx_Oracle.STRING'>, 1, 1, 0, 0, 0), ('CORE', <type 'cx_Oracle.NUMBER'>, 6, 22, 5, 0, 0), ('CORE_TYPE', <type 'cx_Oracle.STRING'>, 1, 1, 0, 0, 0), ('SECTION_NUMBER', <type 'cx_Oracle.NUMBER'>, 3, 22, 2, 0, 0), ('SECTION_TYPE', <type 'cx_Oracle.STRING'>, 2, 2, 0, 0, 1), ('TOP_CM', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, 0, 1), ('BOT_CM', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, 0, 1), ('DEPTH_MBSF', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, 0, 1), ('INOR_C_WT_PCT', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, 0, 1), ('CACO3_WT_PCT', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, 0, 1), ('TOT_C_WT_PCT', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, 0, 1), ('ORG_C_WT_PCT', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, 0, 1), ('NIT_WT_PCT', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, 0, 1), ('SUL_WT_PCT', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, 0, 1), ('H_WT_PCT', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, 0, 1)]

------------------- synonym of view ocd_chem_carb_v created from query
[('LEG', <type 'cx_Oracle.NUMBER'>, 6, 22, 5, 0, 0), ('SITE', <type 'cx_Oracle.NUMBER'>, 7, 22, 6, 0, 0), ('HOLE', <type 'cx_Oracle.STRING'>, 1, 1, 0, 0, 0), ('CORE', <type 'cx_Oracle.NUMBER'>, 6, 22, 5, 0, 0), ('CORE_TYPE', <type 'cx_Oracle.STRING'>, 1, 1, 0, 0, 0), ('SECTION_NUMBER', <type 'cx_Oracle.NUMBER'>, 3, 22, 2, 0, 0), ('SECTION_TYPE', <type 'cx_Oracle.STRING'>, 2, 2, 0, 0, 1), ('TOP_CM', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, -127, 1), ('BOT_CM', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, -127, 1), ('DEPTH_MBSF', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, -127, 1), ('INOR_C_WT_PCT', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, -127, 1), ('CACO3_WT_PCT', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, -127, 1), ('TOT_C_WT_PCT', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, -127, 1), ('ORG_C_WT_PCT', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, -127, 1), ('NIT_WT_PCT', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, -127, 1), ('SUL_WT_PCT', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, -127, 1), ('H_WT_PCT', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, -127, 1)]

------------------- table ocd_chem_carb_table created from view
[('LEG', <type 'cx_Oracle.NUMBER'>, 6, 22, 5, 0, 0), ('SITE', <type 'cx_Oracle.NUMBER'>, 7, 22, 6, 0, 0), ('HOLE', <type 'cx_Oracle.STRING'>, 1, 1, 0, 0, 0), ('CORE', <type 'cx_Oracle.NUMBER'>, 6, 22, 5, 0, 0), ('CORE_TYPE', <type 'cx_Oracle.STRING'>, 1, 1, 0, 0, 0), ('SECTION_NUMBER', <type 'cx_Oracle.NUMBER'>, 3, 22, 2, 0, 0), ('SECTION_TYPE', <type 'cx_Oracle.STRING'>, 2, 2, 0, 0, 1), ('TOP_CM', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, -127, 1), ('BOT_CM', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, -127, 1), ('DEPTH_MBSF', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, -127, 1), ('INOR_C_WT_PCT', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, -127, 1), ('CACO3_WT_PCT', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, -127, 1), ('TOT_C_WT_PCT', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, -127, 1), ('ORG_C_WT_PCT', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, -127, 1), ('NIT_WT_PCT', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, -127, 1), ('SUL_WT_PCT', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, -127, 1), ('H_WT_PCT', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, -127, 1)]

**Comment by fils** [comment]
@tgulacsi @divdat  Rana
   I will be in the air today traveling, but I will try and pull things down prior to leaving and I have free inflight internet.  I will give it all a try.   I did pull down and switch to the issue36 branch and I have built the driver.    I am talking with Pat on slack now, and I think this branch might be working!  I've run some quick tests before I leave that are passing.  Unless I missed something we might have a winner.  

 Pat worked up a REAL test and I about have the go test code ready.    We will still work that up and provide it to you.

Thanks for all the help guys!
Doug

**Comment by fils** [comment]
@tgulacsi @divdat Rana
I've made a test (an ugly test but it works) at  https://gist.github.com/fils/e0f3fc29172d39362213

against branch issue36 all these test pass.  I've not switched back to master yet to see if these fail then.  

I'll try and update this later with better details...  however, this test works against branch issue36.  

Pat do you have any comments on this version of the test.

**Comment by tgulacsi** [comment]
Thanks for all!

I don't have fils.ocd_chem_carb_test_v, but qry3 reproduces the underflow
problem!

Master did produce the issue, and the issue36 modifications fixed it, so
I'd say the issue is fixed.
(At least till a new corner case found...)

---
https://onename.com/tgulacsi

2015-09-16 0:51 GMT+02:00 Douglas Fils <notifications@github.com>:

**Comment by fils** [comment]
@tgulacsi @divdat
I think if you alter "fils" to whatever username you use in the connection string?  Pat would know for sure, perhaps there is something that has to be created in the DB to support that.   I am not the DBA, so I likely should be careful or I will say something stupid.  

I just tried to pull master down but I did something wrong and messed it up.  I will try and pull down master later today and confirm I can get all 6 tests to pass I worked up.    I'm hoping this is it.

I agree, thanks all!

**Comment by divdat** [comment]
Yes, you will have to change any references to "fils." to your oracle user such as "tgulacsi."

I'm glad to hear that this is resolved!  Thanks very much!

**Comment by divdat** [comment]
As a follow up, it is interesting to see however slight, the difference in the "describe" of the queries (using cx_Oracle)

1.  query itself
2.  query to public synonym of the view built from the query
3.  query to view
4.  query to table built from the view

------------------- raw query (to test tables)
[('LEG', <type 'cx_Oracle.NUMBER'>, 6, 22, 5, 0, 0), ('SITE', <type 'cx_Oracle.NUMBER'>, 7, 22, 6, 0, 0), ('HOLE', <type 'cx_Oracle.STRING'>, 1, 1, 0, 0, 0), ('CORE', <type 'cx_Oracle.NUMBER'>, 6, 22, 5, 0, 0), ('CORE_TYPE', <type 'cx_Oracle.STRING'>, 1, 1, 0, 0, 0), ('SECTION_NUMBER', <type 'cx_Oracle.NUMBER'>, 3, 22, 2, 0, 0), ('SECTION_TYPE', <type 'cx_Oracle.STRING'>, 2, 2, 0, 0, 1), ('TOP_CM', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, 0, 1), ('BOT_CM', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, 0, 1), ('DEPTH_MBSF', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, 0, 1), ('INOR_C_WT_PCT', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, 0, 1), ('CACO3_WT_PCT', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, 0, 1), ('TOT_C_WT_PCT', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, 0, 1), ('ORG_C_WT_PCT', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, 0, 1), ('NIT_WT_PCT', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, 0, 1), ('SUL_WT_PCT', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, 0, 1), ('H_WT_PCT', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, 0, 1)]
------------------- public synonym (ocd_chem_carb_test) of view (fils.ocd_chem_carb_test_v) created from query
[('LEG', <type 'cx_Oracle.NUMBER'>, 6, 22, 5, 0, 0), ('SITE', <type 'cx_Oracle.NUMBER'>, 7, 22, 6, 0, 0), ('HOLE', <type 'cx_Oracle.STRING'>, 1, 1, 0, 0, 0), ('CORE', <type 'cx_Oracle.NUMBER'>, 6, 22, 5, 0, 0), ('CORE_TYPE', <type 'cx_Oracle.STRING'>, 1, 1, 0, 0, 0), ('SECTION_NUMBER', <type 'cx_Oracle.NUMBER'>, 3, 22, 2, 0, 0), ('SECTION_TYPE', <type 'cx_Oracle.STRING'>, 2, 2, 0, 0, 1), ('TOP_CM', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, -127, 1), ('BOT_CM', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, -127, 1), ('INOR_C_WT_PCT', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, -127, 1), ('CACO3_WT_PCT', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, -127, 1), ('TOT_C_WT_PCT', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, -127, 1), ('ORG_C_WT_PCT', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, -127, 1), ('NIT_WT_PCT', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, -127, 1), ('SUL_WT_PCT', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, -127, 1), ('H_WT_PCT', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, -127, 1)]
------------------- view (fils.ocd_chem_carb_test_v) created from query
[('LEG', <type 'cx_Oracle.NUMBER'>, 6, 22, 5, 0, 0), ('SITE', <type 'cx_Oracle.NUMBER'>, 7, 22, 6, 0, 0), ('HOLE', <type 'cx_Oracle.STRING'>, 1, 1, 0, 0, 0), ('CORE', <type 'cx_Oracle.NUMBER'>, 6, 22, 5, 0, 0), ('CORE_TYPE', <type 'cx_Oracle.STRING'>, 1, 1, 0, 0, 0), ('SECTION_NUMBER', <type 'cx_Oracle.NUMBER'>, 3, 22, 2, 0, 0), ('SECTION_TYPE', <type 'cx_Oracle.STRING'>, 2, 2, 0, 0, 1), ('TOP_CM', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, -127, 1), ('BOT_CM', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, -127, 1), ('INOR_C_WT_PCT', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, -127, 1), ('CACO3_WT_PCT', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, -127, 1), ('TOT_C_WT_PCT', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, -127, 1), ('ORG_C_WT_PCT', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, -127, 1), ('NIT_WT_PCT', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, -127, 1), ('SUL_WT_PCT', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, -127, 1), ('H_WT_PCT', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, -127, 1)]
------------------- table (fils.ocd_chem_carb_test_table) created from view (fils.ocd_chem_carb_test_v)
[('LEG', <type 'cx_Oracle.NUMBER'>, 6, 22, 5, 0, 0), ('SITE', <type 'cx_Oracle.NUMBER'>, 7, 22, 6, 0, 0), ('HOLE', <type 'cx_Oracle.STRING'>, 1, 1, 0, 0, 0), ('CORE', <type 'cx_Oracle.NUMBER'>, 6, 22, 5, 0, 0), ('CORE_TYPE', <type 'cx_Oracle.STRING'>, 1, 1, 0, 0, 0), ('SECTION_NUMBER', <type 'cx_Oracle.NUMBER'>, 3, 22, 2, 0, 0), ('SECTION_TYPE', <type 'cx_Oracle.STRING'>, 2, 2, 0, 0, 1), ('TOP_CM', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, -127, 1), ('BOT_CM', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, -127, 1), ('INOR_C_WT_PCT', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, -127, 1), ('CACO3_WT_PCT', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, -127, 1), ('TOT_C_WT_PCT', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, -127, 1), ('ORG_C_WT_PCT', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, -127, 1), ('NIT_WT_PCT', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, -127, 1), ('SUL_WT_PCT', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, -127, 1), ('H_WT_PCT', <type 'cx_Oracle.NUMBER'>, 127, 22, 0, -127, 1)]

**Comment by tgulacsi** [comment]
PTAL master branch, it has a new string number type.
